### ISSUE TITLE
bf16_mode finetuning

### ISSUE THREAD

**Issue by leonardledy** [microsoft/aurora#119]
Hi !
I tried to finetune the Air Pollution model using bf16_mode, but it does not seem to work well with the flash attention (at least for me): 
_Error detected in ScaledDotProductFlashAttentionBackward0. No forward pass information available_
_RuntimeError: Function 'ScaledDotProductFlashAttentionBackward0' returned nan values in its 1th output_

It works fine with the previous implementation of autocast though.
Thought you might be interested in the feedback !

**Comment by HP-eter** [comment]
I'm also having issues with `bf16_mode`, mainly getting illegal memory access errors when finetuning the model on new variables. Using autocast triggers the error too when loss.backward() gets executed.

**Comment by wesselb** [comment]
Hey @leonardledy! Unfortunately stability is really an issue.

If you can make it work with @crisbodnar's suggestion and a simple `torch.autocast("cuda", dtype=torch.bfloat16)`, then that would be the optimal, but it really depends on whether you can make it fit into memory.

**Comment by leonardledy** [comment]
Thanks @crisbodnar @wesselb for your suggestions!

Does this mean that flash attention should not even be used with `torch.autocast("cuda", dtype=torch.bfloat16)` ?

**Comment by wesselb** [comment]
@leonardledy FlashAttention cannot be used because it does not support masking, which we use. However, `SDPBackend.CUDNN_ATTENTION` should be compatible.

Autocast with `torch.bfloat16` should be used in conjunction with a better attention kernel.

**Comment by leonardledy** [comment]
Hi @wesselb @crisbodnar !
I experimented a bit with the attention backends and I came to the following conclusions:

- In `bf16_mode`, it seems that for some reason the backend always falls back to flash attention even when specifying another backend using `with sdpa_kernel(SDPBackend.CUDNN_ATTENTION)`. I also tried with EFFICIENT_ATTENTION and MATH. I do not get this problem when I use autocast with `torch.bfloat16`. Maybe the problem comes from my implementation but I thought it was worth mentioning.
- More importantly, the attention backends that use fused kernels (FLASH_ATTENTION, EFFICIENT_ATTENTION and CUDNN_ATTENTION) require query, key and value to be of dimension 4, whereas they are of dimension 5 in Aurora:
```
            q, k, v = map(lambda t: rearrange(t, "(B nW) H N D -> B nW H N D", nW=nW), (q, k, v))
            mask = mask.unsqueeze(1).unsqueeze(0)  # (1, nW, 1, ws, ws)
            x = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=attn_dropout)
            x = rearrange(x, "B nW H N D -> (B nW) H N D")
```
This leads (when using autocast) to `scaled_dot_product_attention` falling back to the classic implementation (MATH), which is not the most efficient.
I tried keeping q, k and v in dimension 4, which enabled the use of EFFICIENT_ATTENTION and led to improvements in both speed and memory usage.

I thus wonder: why did you choose to reshape q, k and v ? Did I miss something ?

**Comment by wesselb** [comment]
Hey @leonardledy!

Using `autocast` with `torch.bfloat16` is much stabler to run than `bf16_mode` and is the preferred option if you can fit it into memory. I will update the docs to state this. `bf16_mode` is an experimental attempt to enable gradient computation on a single A100, but unfortunately seems to suffer from a lot of issues. Perhaps we should discourage its use.

We are also aware of the dimension issue w.r.t. using different attention kernels. I've not yet had the capacity to fix this in the open-source implementation. This is something that can and should be improved.

Thanks for your detailed investigation into the internals of the model. This is massively helpful. :)

**Comment by leonardledy** [comment]
@wesselb thank you for your answer !

If you want I can submit a PR with the little changes I made in order to be able to use fused attention kernels. I also tested it on inference and it seems to work fine
